### PR TITLE
[Refactoring] More strong-typing in IMAP

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

**Comment by chibenwa** [comment]
```
[ca25a34ba497cd01c4de828e7c4e0e020e3900e8] 06:04:46.728 [ERROR] spark.Spark - ignite failed
[ca25a34ba497cd01c4de828e7c4e0e020e3900e8] java.net.BindException: Address already in use
[ca25a34ba497cd01c4de828e7c4e0e020e3900e8] 	at java.base/sun.nio.ch.Net.bind0(Native Method)
```

test this please

File: protocols/imap/src/main/java/org/apache/james/imap/message/request/StoreRequest.java [hunk 1]
```diff
@@ -113,10 +81,10 @@ public String toString() {
         if (isSilent()) {
             builder.append("SILENT ");
         }
-        if (isSignedPlus()) {
+        if (flagsUpdateMode.equals(MessageManager.FlagsUpdateMode.ADD)) {
```
Comment (mbaechler) [inline protocols/imap/src/main/java/org/apache/james/imap/message/request/StoreRequest.java]: comparison for enum is supposed to be `==`

File: protocols/imap/src/main/java/org/apache/james/imap/api/message/FetchData.java [hunk 2]
```diff
@@ -28,6 +28,81 @@
...
+            itemToFetch.add(Item.UID);
+            return this;
+        }
+
+        public Builder fetchModSeq() {
+            itemToFetch.add(Item.MODSEQ);
+            return this;
+        }
+
+        public Builder setChangedSince(long changedSince) {
```
Comment (mbaechler) [inline protocols/imap/src/main/java/org/apache/james/imap/api/message/FetchData.java]: why `set`?

File: protocols/imap/src/main/java/org/apache/james/imap/api/message/FetchData.java [hunk 3]
```diff
@@ -19,236 +19,147 @@
...
+            bodyElements = new HashSet<>();
+            bodyElements.addAll(fetchData.bodyElements);
+            return this;
+        }
 
-    public FetchData setBodyStructure(boolean bodyStructure) {
-        this.bodyStructure = bodyStructure;
-        return this;
-    }
+        public Builder fetch(Item item) {
```
Comment (mbaechler) [inline protocols/imap/src/main/java/org/apache/james/imap/api/message/FetchData.java]: what about using a vararg?

File: protocols/imap/src/main/java/org/apache/james/imap/api/message/FetchData.java [hunk 4]
```diff
@@ -18,237 +18,149 @@
...
-        return bodyStructure;
-    }
+    public static class Builder {
+        private EnumSet<Item> itemToFetch = EnumSet.noneOf(Item.class);
+        private Set<BodyFetchElement> bodyElements = new HashSet<>();
+        private boolean setSeen = false;
+        private long changedSince = -1;
+        private boolean vanished;
+
+        public Builder from(FetchData fetchData) {
```
Comment (mbaechler) [inline protocols/imap/src/main/java/org/apache/james/imap/api/message/FetchData.java]: why it's not a static factory?

File: protocols/imap/src/main/java/org/apache/james/imap/api/message/FetchData.java [hunk 5]
```diff
@@ -18,237 +18,149 @@
...
+            itemToFetch.addAll(Arrays.asList(item));
+            return this;
+        }
 
-    public boolean isEnvelope() {
-        return envelope;
-    }
+        public Builder changedSince(long changedSince) {
+            this.changedSince = changedSince;
+            itemToFetch.add(Item.MODSEQ);
```
Comment (mbaechler) [inline protocols/imap/src/main/java/org/apache/james/imap/api/message/FetchData.java]: you should call the right builder method instead

**Comment by chibenwa** [comment]
```
04:17:52.436 [ERROR] spark.Spark - ignite failed
[ab9cae21ee65227f81f9aa71f2dbfbbc6a32b635] java.net.BindException: Address already in use
```

test this please

File: protocols/imap/src/main/java/org/apache/james/imap/api/message/FetchData.java [hunk 6]
```diff
@@ -73,6 +75,16 @@ public Builder add(BodyFetchElement element, boolean peek) {
             return this;
         }
 
+        private Builder addBodyElements(Collection<BodyFetchElement> elements) {
+            bodyElements.addAll(elements);
+            return this;
+        }
+
+        private Builder setSeen(boolean setSeen) {
```
Comment (mbaechler) [inline protocols/imap/src/main/java/org/apache/james/imap/api/message/FetchData.java]: why `set`?

Comment (chibenwa) [reply protocols/imap/src/main/java/org/apache/james/imap/api/message/FetchData.java]: pff sorry I'm slow today...

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**